File: packages/go-sdk/operations/jobs.go
package operations

import "context"

// JobsService provides job management operations.
type JobsService struct {
	r Requester
}

// NewJobsService creates a new JobsService.
func NewJobsService(r Requester) *JobsService {
	return &JobsService{r: r}
}

// List lists all jobs.
func (s *JobsService) List(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/jobs", query, nil, nil, &result)
	return result, err
}

// Create creates a new job.
func (s *JobsService) Create(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/jobs", nil, nil, body, &result)
	return result, err
}

// Get gets a job by ID.
func (s *JobsService) Get(ctx context.Context, jobID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/jobs/{jobID}", map[string]string{"jobID": jobID}), nil, nil, nil, &result)
	return result, err
}

// Update updates a job.
func (s *JobsService) Update(ctx context.Context, jobID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "PATCH", PathParams("/v1/jobs/{jobID}", map[string]string{"jobID": jobID}), nil, nil, body, &result)
	return result, err
}

// Delete deletes a job.
func (s *JobsService) Delete(ctx context.Context, jobID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/jobs/{jobID}", map[string]string{"jobID": jobID}), nil, nil, nil, nil)
}

// Clone clones a job.
func (s *JobsService) Clone(ctx context.Context, jobID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/jobs/{jobID}/clone", map[string]string{"jobID": jobID}), nil, nil, body, &result)
	return result, err
}

// GetHealth gets the health score for a job.
func (s *JobsService) GetHealth(ctx context.Context, jobID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/jobs/{jobID}/health", map[string]string{"jobID": jobID}), nil, nil, nil, &result)
	return result, err
}

// Trigger triggers a job run.
func (s *JobsService) Trigger(ctx context.Context, jobID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/jobs/{jobID}/trigger", map[string]string{"jobID": jobID}), nil, nil, body, &result)
	return result, err
}

// BulkTrigger triggers multiple runs of a job.
func (s *JobsService) BulkTrigger(ctx context.Context, jobID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/jobs/{jobID}/trigger/bulk", map[string]string{"jobID": jobID}), nil, nil, body, &result)
	return result, err
}

// ListVersions lists job versions.
func (s *JobsService) ListVersions(ctx context.Context, jobID string, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/jobs/{jobID}/versions", map[string]string{"jobID": jobID}), query, nil, nil, &result)
	return result, err
}

// GetVersion gets a specific job version.
func (s *JobsService) GetVersion(ctx context.Context, jobID, versionID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/jobs/{jobID}/versions/{versionID}", map[string]string{"jobID": jobID, "versionID": versionID}), nil, nil, nil, &result)
	return result, err
}

// ListDependencies lists job dependencies.
func (s *JobsService) ListDependencies(ctx context.Context, jobID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/jobs/{jobID}/dependencies", map[string]string{"jobID": jobID}), nil, nil, nil, &result)
	return result, err
}

// CreateDependency creates a job dependency.
func (s *JobsService) CreateDependency(ctx context.Context, jobID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", PathParams("/v1/jobs/{jobID}/dependencies", map[string]string{"jobID": jobID}), nil, nil, body, &result)
	return result, err
}

// DeleteDependency deletes a job dependency.
func (s *JobsService) DeleteDependency(ctx context.Context, jobID, depID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/jobs/{jobID}/dependencies/{depID}", map[string]string{"jobID": jobID, "depID": depID}), nil, nil, nil, nil)
}

// Batch creates jobs in batch.
func (s *JobsService) Batch(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/jobs/batch", nil, nil, body, &result)
	return result, err
}

// BatchDisable disables jobs in batch.
func (s *JobsService) BatchDisable(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/jobs/batch-disable", nil, nil, body, &result)
	return result, err
}

// BatchEnable enables jobs in batch.
func (s *JobsService) BatchEnable(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/jobs/batch-enable", nil, nil, body, &result)
	return result, err
}

File: packages/rust-sdk/src/operations/jobs.rs
use crate::client::StraitClient;
use crate::errors::StraitError;
use crate::http::substitute_path_params;
use std::sync::Arc;

pub struct JobsService {
    client: Arc<StraitClient>,
}

impl JobsService {
    pub fn new(client: Arc<StraitClient>) -> Self {
        Self { client }
    }

    pub async fn list(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/jobs", query, None, None)
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/jobs", None, None, Some(body))
            .await
    }

    pub async fn get(&self, job_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}", &[("jobID", job_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn update(
        &self,
        job_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}", &[("jobID", job_id)]);
        self.client
            .do_request("PATCH", &path, None, None, Some(body))
            .await
    }

    pub async fn delete(&self, job_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}", &[("jobID", job_id)]);
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn clone(
        &self,
        job_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/clone", &[("jobID", job_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn get_health(&self, job_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/health", &[("jobID", job_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn trigger(
        &self,
        job_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/trigger", &[("jobID", job_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn bulk_trigger(
        &self,
        job_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/trigger/bulk", &[("jobID", job_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn list_versions(
        &self,
        job_id: &str,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/versions", &[("jobID", job_id)]);
        self.client
            .do_request("GET", &path, query, None, None)
            .await
    }

    pub async fn get_version(
        &self,
        job_id: &str,
        version_id: &str,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params(
            "/v1/jobs/{jobID}/versions/{versionID}",
            &[("jobID", job_id), ("versionID", version_id)],
        );
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn list_dependencies(&self, job_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/dependencies", &[("jobID", job_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn create_dependency(
        &self,
        job_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/jobs/{jobID}/dependencies", &[("jobID", job_id)]);
        self.client
            .do_request("POST", &path, None, None, Some(body))
            .await
    }

    pub async fn delete_dependency(&self, job_id: &str, dep_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params(
            "/v1/jobs/{jobID}/dependencies/{depID}",
            &[("jobID", job_id), ("depID", dep_id)],
        );
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }

    pub async fn batch(&self, body: serde_json::Value) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/jobs/batch", None, None, Some(body))
            .await
    }

    pub async fn batch_disable(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/jobs/batch-disable", None, None, Some(body))
            .await
    }

    pub async fn batch_enable(
        &self,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/jobs/batch-enable", None, None, Some(body))
            .await
    }
}
